File: scripts/measure-viewcount-change.js
#!/usr/bin/env node
/**
 * Measure YouTube's 24 Aug 2026 view-counting change and write the calibration
 * factor into public/view-count-change.json.
 *
 * From that date a view counts the moment a video begins to play, with no
 * minimum watch time. Long-form previously needed roughly 30 seconds. Every view
 * figure that spans the date is therefore a blend of two metrics, and the
 * revenue estimate (views x RPM) overstates from that day on, because YouTube
 * pays on "engaged views" — the old bar — not the new public count.
 *
 * WHY NOT A SIMPLE BEFORE/AFTER AVERAGE: a channel's daily views are dominated
 * by whatever it uploaded recently, so a whole-channel average conflates the
 * counting change with upload cadence — and August 2026 also has the view and
 * subscriber anomaly running through it. Two controls remove that:
 *
 *   DORMANT, NEVER-ADVERTISED BACK CATALOGUE. Old videos are in flat decay, and
 *   nothing about them changes on 24 Aug except how their views are counted.
 *   Advertised videos are excluded outright: his Walmart re-runs pushed 36,640
 *   views a day through the back catalogue in mid-August against a ~1,100
 *   baseline, which would swamp the signal.
 *
 *   SHORTS AS A CONTROL. Shorts have counted this way since 31 Mar 2025, so they
 *   should show NO step. Dividing the long-form ratio by the Shorts ratio cancels
 *   any channel-wide drift that happens to coincide with the date.
 *
 * The measured noise floor is +/-17% day to day. Resampling the real spread puts
 * the 95% interval on the FACTOR at +/-16% over 7 days either side, +/-11% over
 * 14 and +/-8% over 28 - so a 20-30% step is clear at two weeks but NOT at one.
 * No percentage is published until it clears its own window floor.
 *
 * Usage:  node scripts/measure-viewcount-change.js [--window 14] [--dry]
 */
const fs = require('fs'), zlib = require('zlib'), path = require('path');

const ROOT = path.join(__dirname, '..');
const PUBLIC = path.join(ROOT, 'public');
const SNAP = process.env.SNAPSHOTS_DIR
  || path.join(path.dirname(ROOT), 'jerminaldecline-snapshots', 'snapshots');
const CFG = path.join(PUBLIC, 'view-count-change.json');
const TQ = 'UCfwE_ODI1YTbdjkzuSi1Nag';

const argv = process.argv.slice(2);
const WINDOW = +(argv[argv.indexOf('--window') + 1]) || 14;
const DRY = argv.includes('--dry');

const cfg = JSON.parse(fs.readFileSync(CFG, 'utf8'));
const CHANGE = cfg.effectiveDate;
const data = JSON.parse(fs.readFileSync(path.join(PUBLIC, 'data.json'), 'utf8'));
const adIds = new Set(Object.values(JSON.parse(fs.readFileSync(path.join(PUBLIC, 'ad-videos.json'), 'utf8')).channels)
  .flatMap(c => c.videoIds));

const shift = (d, n) => new Date(Date.parse(d + 'T00:00:00Z') + n * 86400000).toISOString().slice(0, 10);
const PRE_FROM = shift(CHANGE, -WINDOW), POST_TO = shift(CHANGE, WINDOW - 1);

// Dormant = published comfortably before the window opens, so it is in flat decay
// rather than still riding its launch curve.
const DORMANT_BEFORE = shift(PRE_FROM, -21);
const pool = data.videos.filter(v => v.channelId === TQ && !v.unavailable
  && v.publishedAt.slice(0, 10) < DORMANT_BEFORE && !adIds.has(v.id));
const isShort = new Map(pool.map(v => [v.id, !!v.isShort]));

const files = fs.readdirSync(SNAP).filter(f => /-AM\.json\.gz$/.test(f)).sort();
const series = [];
for (const f of files) {
  const day = f.slice(0, 10);
  if (day < shift(PRE_FROM, -1) || day > shift(POST_TO, 1)) continue;
  const j = JSON.parse(zlib.gunzipSync(fs.readFileSync(path.join(SNAP, f))));
  const m = new Map(j.videos.map(v => [v.id, v.views || 0]));
  let L = 0, S = 0, nL = 0, nS = 0;
  for (const v of pool) {
    if (!m.has(v.id)) continue;
    if (isShort.get(v.id)) { S += m.get(v.id); nS++; } else { L += m.get(v.id); nL++; }
  }
  series.push({ day, L, S, nL, nS });
}

// Day-over-day gains, skipping days where catalogue membership moved (a deletion
// mid-window would otherwise read as a negative day).
const gains = [];
for (let i = 1; i < series.length; i++) {
  const a = series[i - 1], b = series[i];
  if (a.nL !== b.nL || a.nS !== b.nS) continue;
  gains.push({ day: b.day, L: b.L - a.L, S: b.S - a.S });
}
const med = a => { const s = [...a].sort((x, y) => x - y); const n = s.length;
  return n ? (n % 2 ? s[(n - 1) / 2] : (s[n / 2 - 1] + s[n / 2]) / 2) : null; };
const slice = (from, to, key) => med(gains.filter(g => g.day >= from && g.day <= to)
  .map(g => g[key]).filter(x => x > 0));

// How many usable days actually exist each side, TRUNCATED TO WHOLE WEEKS.
// Two separate reasons. The daily series has a weekday cycle, so an uneven slice
// of it injects bias that looks exactly like a real step. And the nominal
// --window said nothing about how much data had arrived: run on 25 Aug with one
// post-change day, the old code still recorded windowDays: 14.
const daysIn = (from, to, key) => gains.filter(g => g.day >= from && g.day <= to && g[key] > 0).length;
const availPost = daysIn(CHANGE, POST_TO, 'L');
const availPre  = daysIn(PRE_FROM, shift(CHANGE, -1), 'L');
const WEEKS  = Math.floor(Math.min(availPre, availPost) / 7);
const USABLE = WEEKS * 7;

console.log(`change date ${CHANGE}   requested window +/-${WINDOW}d`);
console.log(`days available: ${availPre} pre, ${availPost} post -> using ${USABLE} each side (${WEEKS} whole week(s))`);

if (USABLE < 7) {
  console.log('');
  console.log('Less than a whole week either side. Nothing written - the site keeps showing');
  console.log('the notice with no percentage, which is the correct state until data exists.');
  process.exit(0);
}

const PRE_A = shift(CHANGE, -USABLE), PRE_B = shift(CHANGE, -1);
const POST_B = shift(CHANGE, USABLE - 1);

const preL = slice(PRE_A, PRE_B, 'L'), postL = slice(CHANGE, POST_B, 'L');
const preS = slice(PRE_A, PRE_B, 'S'), postS = slice(CHANGE, POST_B, 'S');

console.log(`control pool: ${pool.filter(v => !v.isShort).length} long-form, ${pool.filter(v => v.isShort).length} shorts (dormant, never advertised)`);
console.log(`  long-form  ${preL == null ? '-' : Math.round(preL)} -> ${postL == null ? '-' : Math.round(postL)} views/day`);
console.log(`  shorts     ${preS == null ? '-' : Math.round(preS)} -> ${postS == null ? '-' : Math.round(postS)} views/day  (control: expect no step)`);

if (preL == null || postL == null) {
  console.log('');
  console.log('Not enough data either side of the change yet - nothing written.');
  process.exit(0);
}
const rawRatio = postL / preL;
const ctrlRatio = (preS && postS) ? postS / preS : null;
const adjusted = ctrlRatio ? rawRatio / ctrlRatio : rawRatio;
// Share of plays that never met the old watch-time bar: if J plays now count for
// every 1 that used to, then (J-1)/J of them fell short.
const shortfall = adjusted > 1 ? (adjusted - 1) / adjusted : 0;

// The 95% interval on the factor, from resampling the real daily spread:
// +/-16% at 7 days each side, +/-11% at 14, +/-8% at 28. That is closely
// 0.42/sqrt(days), which is what this uses.
//
// The previous gate was `Math.abs(adjusted - 1) >= 0.10` - effect size alone,
// with no test of how much data produced it. At 7 days the noise floor is
// +/-16%, so a fixed 10% bar sat BELOW it: it would have certified noise as a
// finding and published a percentage on the live site.
const noise95 = 0.42 / Math.sqrt(USABLE);
const ctrlDrift = ctrlRatio == null ? null : Math.abs(ctrlRatio - 1);
const enoughData   = USABLE >= 14;                            // two whole weeks minimum
const clearsNoise  = Math.abs(adjusted - 1) > noise95;         // beats THIS window's floor
const controlQuiet = ctrlDrift != null && ctrlDrift <= 0.05;   // Shorts must NOT have moved
const confident = enoughData && clearsNoise && controlQuiet;

console.log('');
console.log(`  long-form step      ${((rawRatio - 1) * 100).toFixed(1)}%`);
if (ctrlRatio !== null) console.log(`  shorts drift        ${((ctrlRatio - 1) * 100).toFixed(1)}%  (removed from the figure below)`);
console.log(`  CALIBRATION FACTOR  ${adjusted.toFixed(3)}  +/-${(noise95 * 100).toFixed(0)}% at 95% on ${USABLE} days`);
console.log(`  => ${(shortfall * 100).toFixed(0)}% of plays never met the old watch-time bar`);
console.log('');

if (confident) {
  console.log('  PUBLISHABLE - the site will show this percentage in the notice.');
} else {
  console.log('  HELD BACK - the site shows the notice with NO percentage because:');
  if (!enoughData)   console.log(`     - only ${USABLE} day(s) each side; 14 is the minimum`);
  if (!clearsNoise)  console.log(`     - the ${(Math.abs(adjusted - 1) * 100).toFixed(1)}% effect is inside this window's +/-${(noise95 * 100).toFixed(0)}% noise floor`);
  if (!controlQuiet) console.log(`     - the Shorts control ${ctrlDrift == null ? 'is unavailable' : 'moved ' + (ctrlDrift * 100).toFixed(1) + '%, so drift is contaminating the comparison'}`);
}

cfg.status = confident ? 'measured' : 'provisional';
cfg.measured = {
  measuredOn: new Date().toISOString().slice(0, 10),
  daysEachSide: USABLE,
  wholeWeeks: WEEKS,
  longFormPrePerDay: Math.round(preL),
  longFormPostPerDay: Math.round(postL),
  shortsPrePerDay: preS == null ? null : Math.round(preS),
  shortsPostPerDay: postS == null ? null : Math.round(postS),
  rawRatio: +rawRatio.toFixed(4),
  controlRatio: ctrlRatio == null ? null : +ctrlRatio.toFixed(4),
  factor: +adjusted.toFixed(4),
  noise95: +noise95.toFixed(4),
  playsBelowOldBar: +shortfall.toFixed(4),
  // The site prints a percentage ONLY when this is true. All three conditions
  // must hold; any one failing keeps the notice qualitative.
  confident,
  heldBackBecause: confident ? null : { enoughData, clearsNoise, controlQuiet },
};
if (DRY) { console.log('\n--dry: nothing written.'); process.exit(0); }
fs.writeFileSync(CFG, JSON.stringify(cfg, null, 2) + '\n');
console.log(`\nwrote ${path.relative(ROOT, CFG)}`);

File: scripts/measure-viewcount-change.py
#!/usr/bin/env python3
"""
Measure YouTube's 24 Aug 2026 view-counting change on the dormant back catalogue.

From that date a view counts from the first frame with no minimum watch time;
long-form previously needed roughly 30 seconds. Shorts have counted this way
since 31 Mar 2025 and act as a control.

WHY THIS SUPERSEDES THE EARLIER (ARCHIVED) VERSION
--------------------------------------------------
The first attempt differenced consecutive SNAPSHOTS and divided by elapsed
hours. That is wrong here, and produced a phantom 1.75x step on 28 Aug 2026.

fetch-data.js refreshes only the last RECENT_WINDOW_DAYS = 60 days of uploads on
a normal run (see scripts/fetch-data.js). Every older video is re-enriched
solely by `--audit`, which runs twice a day. The dormant back catalogue is
therefore 100% audit-refreshed: its stored view counts do not move between
audits however many snapshots are taken. Snapshots are written by the UPDATE
workflow on a different schedule, so a snapshot difference measures "how many
audits fell between these two snapshots", not a rate.

That went unnoticed while both schedules were steady and roughly aligned. When
GitHub's scheduler degraded on 26 Aug 2026 the alignment broke, and the
resulting sampling artefact looked exactly like the effect being measured.

THREE CORRECTIONS
-----------------
1. AUDIT-ALIGNED. YouTube's counters advance continuously; we only OBSERVE them
   at audits. So difference at audit boundaries and divide by true audit-to-audit
   elapsed time. Audit times come from the git log (local, no auth, full history)
   rather than the Actions API, which retains only 90 days.

2. PHASE-MATCHED. Audits land near 03:00 and 14:30 UTC, and the two legs differ
   by ~1.3x: 14:30->03:00 reads ~890 views/24h, 03:00->14:30 reads ~700. That is
   a real diurnal cycle, and it survives audit alignment. Early legs are
   therefore compared only with early legs.

3. CLOCK-QUALITY FILTERED. Legs outside 10.0-13.5h cover unequal phase mixes and
   cannot be matched, so they are dropped - and the count of drops is reported,
   because during a scheduler outage that is most of the window and the honest
   answer is "not measurable yet".

Usage:
    python scripts/measure-viewcount-change.py            # report
    python scripts/measure-viewcount-change.py --clock    # clock health only
"""
import os, gzip, json, io, sys, datetime, statistics, subprocess

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PUBLIC = os.path.join(ROOT, 'public')
SNAP = os.environ.get('SNAPSHOTS_DIR') or os.path.join(
    os.path.dirname(ROOT), 'jerminaldecline-snapshots', 'snapshots')

TQ = 'UCfwE_ODI1YTbdjkzuSi1Nag'
CHANGE = '2026-08-24'
DORMANT_BEFORE = '2026-06-25'      # comfortably outside the 60-day refresh window
LEG_MIN, LEG_MAX = 10.0, 13.5      # a well-formed half-day audit leg
SNAP_LAG_MAX = 14.0                # hours; beyond this no snapshot reflects the audit
CLOCK_ONLY = '--clock' in sys.argv


def audit_times():
    """Audit run times from the git log. An audit that changed nothing leaves no
    commit, but then the data did not move either, so nothing is lost."""
    out = subprocess.run(
        ['git', 'log', '--format=%ad', '--date=iso-strict', 'origin/main',
         '--grep=daily audit', '-400'],
        capture_output=True, text=True, cwd=ROOT)
    ts = []
    for line in out.stdout.strip().split('\n'):
        line = line.strip()
        if not line:
            continue
        try:
            ts.append(datetime.datetime.fromisoformat(line))
        except ValueError:
            pass
    return sorted(ts)


def build_pool():
    data = json.load(io.open(os.path.join(PUBLIC, 'data.json'), encoding='utf-8'))
    ads = set()
    for c in json.load(io.open(os.path.join(PUBLIC, 'ad-videos.json'),
                               encoding='utf-8'))['channels'].values():
        ads.update(c.get('videoIds', []))
    return {v['id']: bool(v.get('isShort')) for v in data['videos']
            if v.get('channelId') == TQ and not v.get('unavailable')
            and v.get('publishedAt', '')[:10] < DORMANT_BEFORE
            and v['id'] not in ads}


def read_snapshots(pool, since):
    snaps = []
    for f in sorted(x for x in os.listdir(SNAP)
                    if x.endswith('.json.gz') and x[:10] >= since):
        try:
            j = json.loads(gzip.open(os.path.join(SNAP, f)).read().decode('utf-8'))
        except Exception:
            continue
        t = datetime.datetime.fromisoformat(j['meta']['lastUpdated'].replace('Z', '+00:00'))
        L = S = 0
        for v in j['videos']:
            sid = pool.get(v['id'])
            if sid is None:
                continue
            n = v.get('views') or 0
            if sid:
                S += n
            else:
                L += n
        snaps.append({'t': t, 'L': L, 'S': S})
    return snaps


def legs(audits, snaps):
    """Each audit's values, read from the first snapshot at or after it."""
    obs = []
    for a in audits:
        nxt = next((s for s in snaps if s['t'] >= a), None)
        if not nxt or (nxt['t'] - a).total_seconds() / 3600 > SNAP_LAG_MAX:
            continue
        obs.append({'t': a, 'L': nxt['L'], 'S': nxt['S']})
    # Two audits resolving to the SAME snapshot carry identical values; keep one.
    ded = []
    for o in obs:
        if ded and o['L'] == ded[-1]['L'] and o['S'] == ded[-1]['S']:
            continue
        ded.append(o)

    good, dropped = [], []
    for i in range(1, len(ded)):
        a, b = ded[i - 1], ded[i]
        h = (b['t'] - a['t']).total_seconds() / 3600
        rec = {'t': b['t'], 'h': h,
               'L': (b['L'] - a['L']) / h * 24, 'S': (b['S'] - a['S']) / h * 24,
               'leg': 'early' if b['t'].hour < 9 else 'late'}
        (good if LEG_MIN <= h <= LEG_MAX else dropped).append(rec)
    return good, dropped


def main():
    audits = audit_times()
    if len(audits) < 10:
        print('Could not read audit times from the git log (need origin/main fetched).')
        return 1
    pool = build_pool()
    nL = sum(1 for s in pool.values() if not s)
    nS = sum(1 for s in pool.values() if s)
    since = (min(audits) - datetime.timedelta(days=2)).strftime('%Y-%m-%d')
    snaps = read_snapshots(pool, max(since, '2026-07-01'))
    good, dropped = legs(audits, snaps)

    print('audits found      : %d   (%s -> %s)' % (
        len(audits), audits[0].strftime('%d %b %H:%M'), audits[-1].strftime('%d %b %H:%M')))
    print('pool              : %d long-form, %d shorts (all audit-refreshed)' % (nL, nS))
    print('usable legs       : %d      dropped for irregular spacing: %d' % (len(good), len(dropped)))

    # ---- clock health ------------------------------------------------------
    recent = [r for r in (good + dropped) if r['t'] > audits[-1] - datetime.timedelta(days=7)]
    bad = [r for r in recent if not (LEG_MIN <= r['h'] <= LEG_MAX)]
    print('\nCLOCK HEALTH (last 7 days)')
    print('  legs %d, irregular %d (%.0f%%)' % (len(recent), len(bad),
                                                100 * len(bad) / len(recent) if recent else 0))
    # 20%, not a third. During the clean run to 26 Aug 2026 irregular legs were
    # ~5%; at 33% the sampling is already degraded enough to fabricate a step,
    # which is exactly what happened.
    frac = len(bad) / len(recent) if recent else 0
    if frac > 0.20:
        print('  DEGRADED - the audit schedule is drifting, so most legs cannot be')
        print('  phase-matched. Treat any factor below as provisional; the 14-day')
        print('  count restarts when spacing returns to ~12.6h / ~11.4h.')
    elif frac > 0.05:
        print('  marginal - some drift, watch it')
    else:
        print('  steady')
    if CLOCK_ONLY:
        return 0

    # ---- measurement -------------------------------------------------------
    med = statistics.median
    print('\nAUDIT-ALIGNED, PHASE-MATCHED  (change date %s)' % CHANGE)
    results = []
    for leg in ('early', 'late'):
        pre = [r for r in good if r['leg'] == leg and r['t'].strftime('%Y-%m-%d') < CHANGE]
        post = [r for r in good if r['leg'] == leg and r['t'].strftime('%Y-%m-%d') >= CHANGE]
        label = 'early (->03h)' if leg == 'early' else 'late  (->14h)'
        if len(pre) < 3 or len(post) < 2:
            print('  %s : pre n=%d post n=%d - too few to compare' % (label, len(pre), len(post)))
            continue
        pL, pS = med([r['L'] for r in pre]), med([r['S'] for r in pre])
        qL, qS = med([r['L'] for r in post]), med([r['S'] for r in post])
        cv = statistics.pstdev([r['L'] for r in pre]) / pL if pL else float('nan')
        did = (qL / pL) / (qS / pS) if pS and qS else float('nan')
        results.append(did)
        print('  %s' % label)
        print('    pre  n=%2d : long %6.0f/24h   shorts %6.0f/24h   (baseline CV %.0f%%)'
              % (len(pre), pL, pS, 100 * cv))
        print('    post n=%2d : long %6.0f/24h   shorts %6.0f/24h' % (len(post), qL, qS))
        print('    long %.3fx   shorts %.3fx   difference-in-differences %.3fx'
              % (qL / pL, qS / pS, did))

    if results:
        print('\n  factor across legs: %s' % ' / '.join('%.2fx' % r for r in results))
        print('  Shorts cannot be affected by the counting change - any move there is')
        print('  contamination, which is what the difference-in-differences removes.')
    post_days = sorted({r['t'].strftime('%d %b') for r in good
                        if r['t'].strftime('%Y-%m-%d') >= CHANGE})
    print('\n  post-change days with a usable leg: %s' % (', '.join(post_days) or 'none'))
    return 0


if __name__ == '__main__':
    sys.exit(main())
